File: Producto Escalar 25200/producto_escalar.c
#include <stdio.h>
#include <stdlib.h>
#include <mpi.h>

#define TAMANIO 25200

// Función para llenar un arreglo con valores
void llenar_matriz(int matriz[], int tamanio) {
    for (int i = 0; i < tamanio; i++) {
        matriz[i] = i + 1; // Llena con valores consecutivos, puedes cambiar esta lógica si es necesario
    }
}

// Función para calcular el producto vectorial parcial
long prod_vectorial(int v1[], int v2[], int a, int b) {
    long suma = 0;
    for (int i = a; i < b; i++) {
        suma += v1[i] * v2[i];
    }
    return suma;
}

int main(int argc, char **argv) {
    int arre1[TAMANIO];
    int arre2[TAMANIO];

    // Llenar las matrices
    llenar_matriz(arre1, TAMANIO);
    llenar_matriz(arre2, TAMANIO);

    int dest = 0; // Proceso que recibirá los resultados parciales
    int tag = 0;  // Etiqueta de los mensajes
    int tama = TAMANIO;

    double start_time, end_time, elapsed_time;

    MPI_Status status;

    /* Declaración de variables */
    int ID_Proceso, TotalProcesos, cant_pasos;
    long long resultado_t; long resultado_i;
    int nva_a, nva_b;

    /* Inicialización de MPI */
    MPI_Init(&argc, &argv);
    MPI_Comm_rank(MPI_COMM_WORLD, &ID_Proceso);
    MPI_Comm_size(MPI_COMM_WORLD, &TotalProcesos);

    // Inicio del temporizador
    start_time = MPI_Wtime();

    // División del trabajo
    cant_pasos = tama / TotalProcesos;

    nva_a = ID_Proceso * cant_pasos;
    nva_b = (ID_Proceso + 1) * cant_pasos;

    // Cálculo del producto vectorial parcial
    resultado_i = prod_vectorial(arre1, arre2, nva_a, nva_b);

    printf("Soy el proceso %d y mi suma parcial es %li\n", ID_Proceso, resultado_i);

    if (ID_Proceso == 0) {
        resultado_t = resultado_i;
        for (int i = 1; i < TotalProcesos; i++) {
            MPI_Recv(&resultado_i, 1, MPI_LONG, i, tag, MPI_COMM_WORLD, &status);
            resultado_t += resultado_i;
        }
        printf("\nEl resultado total es %lld\n", resultado_t);
    } else {
        MPI_Send(&resultado_i, 1, MPI_LONG, dest, tag, MPI_COMM_WORLD);
    }

    // Fin del temporizador
    end_time = MPI_Wtime();

    elapsed_time = end_time - start_time;

    // Reducir para obtener el tiempo máximo
    double max_time;
    MPI_Reduce(&elapsed_time, &max_time, 1, MPI_DOUBLE, MPI_MAX, 0, MPI_COMM_WORLD);

    if (ID_Proceso == 0) {
        printf("Tiempo total de ejecución = %f segundos\n", max_time);
    }

    MPI_Finalize();

    return 0;
}

File: Producto Escalar 25200/producto_escalar.py
from mpi4py import MPI
import numpy as np
import time

# Definir el tamaño de los arreglos
TAMANIO = 25200

# Función para llenar un arreglo con valores
def llenar_matriz(tamanio):
    return np.arange(1, tamanio + 1)

# Función para calcular el producto vectorial parcial
def prod_vectorial(v1, v2, a, b):
    suma = 0
    for i in range(a, b):
        suma += v1[i] * v2[i]
    return suma

def main():
    # Inicializar MPI
    comm = MPI.COMM_WORLD
    ID_Proceso = comm.Get_rank()
    TotalProcesos = comm.Get_size()

    # Crear y llenar las matrices
    arre1 = llenar_matriz(TAMANIO)
    arre2 = llenar_matriz(TAMANIO)

    # División del trabajo
    cant_pasos = TAMANIO // TotalProcesos
    nva_a = ID_Proceso * cant_pasos
    nva_b = (ID_Proceso + 1) * cant_pasos

    # Calcular el producto vectorial parcial
    start_time = MPI.Wtime()
    resultado_i = prod_vectorial(arre1, arre2, nva_a, nva_b)

    # Reducir los resultados parciales al proceso 0
    resultado_t = comm.reduce(resultado_i, op=MPI.SUM, root=0)

    # Obtener el tiempo máximo
    end_time = MPI.Wtime()
    elapsed_time = end_time - start_time
    max_time = comm.reduce(elapsed_time, op=MPI.MAX, root=0)

    if ID_Proceso == 0:
        print(f'{max_time}, ', end="")

if __name__ == "__main__":
    main()
